Automattic/newspack-plugin
# Build pipeline, Murielized Checkbox, and Subscriptions Wizard stub

### All Submissions:






### Changes proposed in this Pull Request:



Closes #22.
Closes #16.

This PR sets up the build pipeline for the wizards. A "wizard" is a self-contained admin-screen feature, and each wizard is its own single-page app. The modularity and independence of the wizards should make it easier to develop wizards in parallel and will prevent overhead from trying to coordinate a bunch of wizards in one master single-page app.

#### Webpack

The webpack config is heavily based on the [newspack-blocks](https://github.com/Automattic/newspack-blocks/blob/master/webpack.config.js) webpack config, but with some custom stuff to make it work better for components and wizards.

Within the `src` directory there are two directories: `components` and `wizards`.

The webpack slurps up all of the component css files and builds them into one master `components.css` file that gets enqueued automatically for each wizard.

The webpack builds JS and CSS files for each wizard, named `[name].js` and [name]-style.css` respectively. These can be enqueued as needed in the wizard's PHP class.

All webpack output goes into the `dist` folder.

#### Creating components

Examine the `components/checkboxInput` directory to see a working example.
1. Create a new directory in the `src/components` directory.
2. Add `index.js` and `style.scss` files.
3. In `index.js` make a "wrapper" component for an existing Gutenberg component if possible, adding a class for styling and making any other tweaks we need to the behavior.
4. In `style.scss`, add styling to make the component match the Muriel design styles.

#### Creating wizards

Examine the `assets/src/wizards/subscriptions` directory and `includes/wizards/class-subscriptions-wizard.php` file to see a working example.
1. Create a new directory in the `src/wizards` directory.
2. Add `index.js` and `style.scss` files.
3. Build/style the wizard frontend.
4. Create a new PHP class extending the Wizard class. It will automatically create a page for the wizard to live on. You can add ajax endpoints or other logic to the PHP class to form the backend of the wizard. Enqueue any wizard scripts and styles here, following the example in the Subscriptions Wizard class.

### How to test the changes in this Pull Request:

1. `npm install` then `npm run build:webpack`
2. In WP Admin, you should see a Newspack section with a Subscriptions sub-section.
3. Go to Subscriptions and test the Muriel-designed checkbox example there:
<img width="273" alt="Screen Shot 2019-04-30 at 10 25 19 AM" src="https://user-images.githubusercontent.com/7317227/56982946-4e260a80-6b37-11e9-9ee0-8f815ca32dd9.png">


### Other information:
